pkg/kor/finalizers.go

```diff
@@ -0,0 +1,128 @@
+	pendingDeletionResources := make(map[string]map[schema.GroupVersionResource][]string) //map[namespace]map[gvr][]resourceNames
+
+	for _, apiResourceList := range resourceTypes {
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			return pendingDeletionResources, err
+		}
+
+		for _, resourceType := range apiResourceList.APIResources {
+
+			if slices.Contains(resourceType.Verbs, "list") {
+
+				gvr := gv.WithResource(resourceType.Name)
+				resourceList, err := dynamicClient.
+					Resource(gvr).
+					Namespace(metav1.NamespaceAll).
+					List(context.TODO(), metav1.ListOptions{})
+				if err != nil {
+					fmt.Printf("Error listing resources for GVR %s: %v\n", apiResourceList.GroupVersion, err)
+					continue
+				}
+				for _, item := range resourceList.Items {
+					labels := item.GetLabels()
+					if labels["kor/used"] == "true" {
+						continue
+					}
+
+					// Check for excluded labels
+					if excluded, _ := HasExcludedLabel(labels, filterOpts.ExcludeLabels); excluded {
+						continue
+					}
+
+					// Check age criteria
+					if included, _ := HasIncludedAge(item.GetCreationTimestamp(), filterOpts); !included {
+						continue
+					}
+					if CheckFinalizers(item.GetFinalizers(), item.GetDeletionTimestamp()) {
+						if pendingDeletionResources[item.GetNamespace()] == nil {
+							pendingDeletionResources[item.GetNamespace()] = make(map[schema.GroupVersionResource][]string)
+						}
+						pendingDeletionResources[item.GetNamespace()][gvr] = append(pendingDeletionResources[item.GetNamespace()][gvr], item.GetName())
+					}
+				}
+			}
+		}
+	}
+	return pendingDeletionResources, nil
+}
+
+func getResourcesWithFinalizersPendingDeletion(clientset kubernetes.Interface, dynamicClient dynamic.Interface, filterOpts *FilterOptions) (map[string]map[schema.GroupVersionResource][]string, error) {
+	// Use the discovery client to fetch API resources
+	resourceTypes, err := clientset.Discovery().ServerPreferredNamespacedResources()
+	if err != nil {
+		fmt.Printf("Error fetching server resources: %v\n", err)
+		os.Exit(1)
+	}
+
+	return retrievePendingDeletionResources(resourceTypes, dynamicClient, filterOpts)
+}
+
+func GetUnusedfinalizers(includeExcludeLists IncludeExcludeLists, filterOpts *FilterOptions, clientset kubernetes.Interface, dynamicClient *dynamic.DynamicClient, outputFormat string, opts Opts) (string, error) {
+	var outputBuffer bytes.Buffer
+	namespaces := SetNamespaceList(includeExcludeLists, clientset)
+	response := make(map[string]map[string][]string)
+	pendingDeletionDiffs, err := getResourcesWithFinalizersPendingDeletion(clientset, dynamicClient, filterOpts)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to process resources waiting for finalizers: %v\n", err)
+	}
+
+	allDiffs := make(map[string][]string)
+
+	for namespace, resourceType := range pendingDeletionDiffs {
+		if slices.Contains(namespaces, namespace) {
+			for gvr, resourceDiff := range resourceType {
```

Comment:
What is gvr?

Reply:
schema of group version resource, resource stands for the name of the resource in the same format of the api endpoint which is lowercase plural.
for example: Kind- `Pod`, Resource-`pods`